accession-commons-mongodb/src/main/java/uk/ac/ebi/ampt2d/commons/accession/persistence/mongodb/repository/BasicMongoDbAccessionedCustomRepositoryImpl.java

```diff
@@ -0,0 +1,119 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package uk.ac.ebi.ampt2d.commons.accession.persistence.mongodb.repository;
+
+import com.mongodb.BulkWriteError;
+import org.springframework.data.mongodb.BulkOperationException;
+import org.springframework.data.mongodb.core.BulkOperations;
+import org.springframework.data.mongodb.core.MongoTemplate;
+import uk.ac.ebi.ampt2d.commons.accession.core.SaveResponse;
+import uk.ac.ebi.ampt2d.commons.accession.persistence.IAccessionedObjectCustomRepository;
+import uk.ac.ebi.ampt2d.commons.accession.persistence.mongodb.document.AccessionedDocument;
+
+import java.io.Serializable;
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public abstract class BasicMongoDbAccessionedCustomRepositoryImpl<
+        ACCESSION extends Serializable,
+        DOCUMENT extends AccessionedDocument<ACCESSION>>
+        implements IAccessionedObjectCustomRepository<ACCESSION, DOCUMENT> {
+
+    private final Class<DOCUMENT> clazz;
+    private final MongoTemplate mongoTemplate;
+
+    public BasicMongoDbAccessionedCustomRepositoryImpl(Class<DOCUMENT> clazz, MongoTemplate mongoTemplate) {
+        this.clazz = clazz;
+        this.mongoTemplate = mongoTemplate;
+    }
+
+    @Override
+    public SaveResponse<ACCESSION> insert(List<DOCUMENT> documents) {
+        checkHashUniqueness(documents);
+        putAuditSaveDate(documents);
+        final BulkOperations insert = mongoTemplate.bulkOps(BulkOperations.BulkMode.UNORDERED, clazz)
+                .insert(new ArrayList<>(documents));
+        final Set<String> duplicatedHash = new HashSet<>();
+
+        try {
+            insert.execute();
+        } catch (BulkOperationException e) {
+            e.getErrors().forEach(error -> {
+                String duplicatedId = parseIdDuplicateKey(error).orElseThrow(() -> e);
+                duplicatedHash.add(duplicatedId);
+            });
+        }
+        return generateSaveResponse(documents, duplicatedHash);
+    }
+
+    private void checkHashUniqueness(Collection<DOCUMENT> documents) {
+        final Set<String> duplicatedHash = new HashSet<>();
+        documents.forEach(document -> {
+            if (duplicatedHash.contains(document.getHashedMessage())) {
+                throw new RuntimeException("Duplicated hash in mongodb insert batch");
+            }
+            duplicatedHash.add(document.getHashedMessage());
+        });
+    }
+
+    /**
+     * Unfortunately we need to set this manually when using a bulk operation.
+     * @param documents
+     */
+    private void putAuditSaveDate(Iterable<DOCUMENT> documents) {
```

Comment:
`setAuditCreatedDate` because that is the actual attribute and annotation (also `put` is not that usual in English).